subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/AbstractVirtualFileSystem.java

```diff
@@ -45,31 +49,58 @@ public Optional<MetadataSnapshot> findMetadata(String absolutePath) {
     }
 
     @Override
-    public void store(String absolutePath, FileSystemLocationSnapshot snapshot) {
-        rootReference.update(root -> updateNotifyingListeners(diffListener -> root.store(absolutePath, snapshot, diffListener)));
+    public FileSystemLocationSnapshot store(String absolutePath, Supplier<FileSystemLocationSnapshot> snapshotSupplier) {
+        long versionBefore = versionHierarchyRoot.getVersionFor(absolutePath);
+        FileSystemLocationSnapshot snapshot = snapshotSupplier.get();
+        updateCheckingVersion(absolutePath, versionBefore, snapshot);
+        return snapshot;
+    }
+
+    @Override
+    public FileSystemLocationSnapshot store(String baseLocation, StoringAction storingAction) {
+        long versionBefore = versionHierarchyRoot.getVersionFor(baseLocation);
+        return storingAction.snapshot(snapshot -> updateCheckingVersion(snapshot.getAbsolutePath(), versionBefore, snapshot));
+    }
+
+    private void updateCheckingVersion(String absolutePath, long versionBefore, FileSystemLocationSnapshot snapshot) {
+        long versionAfter = versionHierarchyRoot.getVersionFor(absolutePath);
+        // Only update VFS if no changes happened in between
+        if (versionBefore == versionAfter) {
+            rootReference.update(root -> updateNotifyingListeners(diffListener -> root.store(absolutePath, snapshot, diffListener)));
+        } else {
+            LOGGER.debug("Changes to the virtual file system happened while snapshotting '{}', not storing resulting snapshot", absolutePath);
+        }
     }
 
     @Override
     public void invalidate(Iterable<String> locations) {
         LOGGER.debug("Invalidating VFS paths: {}", locations);
         rootReference.update(root -> {
             SnapshotHierarchy result = root;
+            VersionHierarchyRoot newVersionHierarchyRoot = versionHierarchyRoot;
+            long nextVersion = ++currentVersion;
```

Comment:
I still don't understand why we need a separate `long` to keep track of the version. The version in the root of `versionHierarchyRoot` will by definition always reflect the same number, no? Because when we update the version of any point in the hierarchy, we'll always also update its ancestry, too. Or am I missing something?

Reply:
I removed that and only keep the version in the root.